Detect file name case issues for import

arBmind opened the pull request:
This fixes #9786

andrewrk commented:
Can you share what the user experience is like now when this issue occurs?

arBmind commented:
stage1:
```
PS C:\C\Zig\path_demo\src> C:\C\Zig\repo\stage1\bin\zig.exe test main.zig
.\main.zig:6:17: error: import of 'My_types.zig' mismatches real path: 'C:\C\Zig\path_demo\src\my_types.zig'
const MyType3 = @import("My_types.zig").MyType; // capital M
                ^
```

stage2:
```
PS C:\C\Zig\path_demo\src> zig test main.zig                                                                      
warning(module): real path C:\C\Zig\path_demo\src\my_types.zig != imported path My_types.zig
main.zig:6:25: error: unable to load 'My_types.zig': ErrorImportPathMismatch
const MyType3 = @import("My_types.zig").MyType; // capital M
                        ^
```

Vexu commented on src/Module.zig:
```diff
@@ -1848,6 +1848,17 @@ pub const File = struct {
+            log.warn("real path {s} != imported path {s}", .{
+                real_path, resolved_path,
+            });
+            return error.ErrorImportPathMismatch;
```
Redundant "error"
```suggestion
            return error.ImportPathMismatch;
```

arBmind replied:
Thx for pointing this out. I blindly copied it over from the C enum.

Vexu commented on src/stage1/ir.cpp:
```diff
@@ -17275,6 +17275,10 @@ static Stage1AirInst *ir_analyze_instruction_import(IrAnalyze *ira, Stage1ZirIns
             return ira->codegen->invalid_inst_gen;
+        } else if (err == ErrorImportPathMismatch) {
+            ir_add_error_node(ira, source_node,
+                    buf_sprintf("import of '%s' mismatches real path: '%s'", buf_ptr(import_target_path), buf_ptr(&full_path)));
```
I think this error is too vague, it should mention the likely cause of case mismatch (or truncated leading spaces?) and should not mention `real path` since that doesn't seem to be what the platform this is most likely to occur on (windows) calls it.

arBmind replied:
I had a hard time to get a good message here.
For now I settled with:
`import of '%s' mismatches actual file path: '%s'. Probably wrong capitalisation.`

andrewrk reviewed (changes requested):
```
PS C:\C\Zig\path_demo\src> C:\C\Zig\repo\stage2\bin\zig.exe test main.zig                                                                      
warning(module): real path C:\C\Zig\path_demo\src\my_types.zig != imported path My_types.zig
main.zig:6:25: error: unable to load 'My_types.zig': ErrorImportPathMismatch
const MyType3 = @import("My_types.zig").MyType; // capital M
                        ^
```

I don't like the warning. I think the compile error should be the only message when this occurs. I agree with @Vexu to avoid using the phrase "realpath" since it's not a cross-platform concept, however I disagree that Zig should guess the likely cause of mismatch, for the same reason I think Zig should not guess which identifiers the user meant to use when a wrong identifier is found ("unknown identifier 'foobra', did you mean 'foobar'?"). Reason being that chance of a wrong guess is high, extremely misleading, and destroys trust in the tooling. I want Zig to only report accurate diagnostics.

Wording suggestion:
```
PS C:\C\Zig\path_demo\src> C:\C\Zig\repo\stage2\bin\zig.exe test main.zig
main.zig:6:25: error: unable to load 'My_types.zig': canonical file system name is 'my_types.zig'
const MyType3 = @import("My_types.zig").MyType; // capital M
                        ^
```

arBmind commented:
@andrewrk I can relate to your arguments and agree with the goal. This was not easy to achieve, as we cannot easily return data with the error codes.
For the stage1 we already had the path. I changed the message, but did not bother to shorten the file paths here.
For stage2 we have two places.
1. File-GetSource uses the source of File to return the real path. I also added logic to remove the path if they are equal. This should give us nice messages for 99% of all use cases.
2. For embedFile there was no convenient way. So I created a EmbedFileResult struct like ImportFileResult next door. But I use it to return the real_path in case of an error.

As these are my first steps in Zig, I might have messed up some ideas. Feedback is always welcome.

Vexu commented on src/Compilation.zig:
```diff
@@ -3475,17 +3475,43 @@ fn reportRetryableAstGenError(
-            file.sub_file_path, @errorName(err),
-        });
+    const err_msg = switch (err) {
+        error.ImportPathMismatch => blk: {
```
I haven't touched this part of the code in a while but to me it seems like this belongs in `Sema.zirImport`.

arBmind replied:
I have not moved this code. The scope of the function seems correct, as it is supposed to report this kind of error.

Should I change anything here?

Vexu commented on src/Sema.zig:
```diff
@@ -8720,25 +8720,29 @@ fn zirEmbedFile(sema: *Sema, block: *Block, inst: Zir.Inst.Index) CompileError!A
             return sema.fail(block, operand_src, "unable to open '{s}': {s}", .{ name, @errorName(err) });
         },
     };
+    if (result.embed_file) |embed_file| {
```
Style nit but I prefer using `orelese` in cases like this to avoid extra indentation:
```zig
const embed_file = result.embed_file orelse {
    defer mod.gpa.free(result.real_path);
    return sema.fail(block, operand_src, "unable to load '{s}': canonical file system name is '{s}'", .{ name, result.real_path });
};
```

arBmind commented on src/Module.zig:
```diff
@@ -4339,6 +4362,12 @@ pub fn embedFile(mod: *Module, cur_file: *File, rel_file_path: []const u8) !*Emb
     const bytes = try file.readToEndAllocOptions(gpa, std.math.maxInt(u32), size_usize, 1, 0);
     errdefer gpa.free(bytes);
 
+    const real_path = try std.fs.realpathAlloc(gpa, resolved_path);
+    if (!mem.endsWith(u8, real_path, sub_file_path)) {
+        return EmbedFileResult{ .real_path = real_path };
+    }
+    defer gpa.free(real_path);
```
Intuitively I thought the same. But…

We want to successfully load the file first before checking for the path naming issue. If we reverse the order even missing files have to be handled through this error, but this would require a different error message. Reading the file might still go wrong because of some missing privileges. So we would have to do it twice.

Doing the extra work of loading the file should not be an issue here, as most errors will occur only after loading the file during the processing of it.

Vexu replied:
Not sure I understand what you're saying.

> We want to successfully load the file first before checking for the path naming issue. If we reverse the order even missing files have to be handled through this error, but this would require a different error message. Reading the file might still go wrong because of some missing privileges. So we would have to do it twice.

Why? Does `realpath` not report all the same errors as `open`?

arBmind replied:
I don't know. I would guess this is depending on the actual platform and implementation of `real_path` for that platform. There are already multiple version in the zig standard library. I had to choose one that is supported on all compiler platforms. As I don't own all of them I tried to make the least amount of assumptions about them.

arBmind replied:
> Why? Does `realpath` not report all the same errors as `open`?

Looking a bit deeper it seems that different operating systems use different strategies for real path here.
The common platforms like posix and windows open the file and use the file handle to retrieve the real path.
But WASI is currently a very simple emulation using just resolve.
The libc implementation is also free to do whatever it likes to do.

So it seems the file is opened twice for some platforms, because this is the `realpath` is implemented for most platforms (unlike eg. getFdPath).
This concerns me more than the order of operations. But I guess we don't want to optimize on the os level in the compiler code, yet.

Checking after loading the file should no be an issue. If we get an error here, the compilation was fast anyways, as we did not parse and analyze the contents.
So why bother with that?

arBmind commented on src/Module.zig:
```diff
@@ -1848,6 +1848,15 @@ pub const File = struct {
         if (amt != stat.size)
             return error.UnexpectedEndOfFile;
 
+        const resolved_path = try file.fullPath(gpa);
+        defer gpa.free(resolved_path);
+        const real_path = try std.fs.realpathAlloc(gpa, resolved_path);
+        if (!mem.endsWith(u8, real_path, file.sub_file_path)) {
+            file.source = @ptrCast([:0]const u8, real_path);
+            return error.ImportPathMismatch;
+        }
+        defer gpa.free(real_path);
```
Hi @Vexu, thank you for clarifying your point here.

My reasoning was as follows:

1. Unlike `embedFile` the `importFile` method does not actually load the file immediately. This is deferred to the thread that handles the parsing of the zig file. Therefore I assume we should do the least amount of (IO) work on the importing thread.
2. We want to load the file first before checking for the path naming issue. See my comment on `embedFile` below.

Both reasons lead me to implement the check only when the `File` is loaded. Please correct me if my assumptions are wrong here.

Vexu replied:
>1. Unlike `embedFile` the `importFile` method does not actually load the file immediately. This is deferred to the thread that handles the parsing of the zig file. Therefore I assume we should do the least amount of (IO) work on the importing thread.

By the time we get to `Sema.importFile` all relevant files should already have been loaded by AstGen.

>2. We want to load the file first before checking for the path naming issue. See my comment on `embedFile` below.

I'm not sure why we would want to do that but it doesn't apply here anyways since file would already be loaded.

arBmind replied:
Sorry for being stubborn, but I have a hard time to follow your thoughts.

From analysing the code it looks like this:
* In `Module` we have `fn astGenFile`. This method seems to be the only place an imported zig file is loaded, after checking for compilation cache and so on.
* This method is invoked from `fn workerAstGenFile` on `Compilation`, which delegates error handling to `fn reportRetryableAstGenError`.

From the other side:
* In `Sema` we have `fn zirImport` that calls `fn importFile` from `Module` which creates an "unloaded" File instance and then creates a `file.root_decl` that is the referenced thing that is returned. But no actual file loading is started here.

For sure we can move the `real_path` check to the `zirImport` side. But this would also trigger the check even if we have already cached the compiled binary for this file. I don't the see advantage.

arBmind commented:
It seems this PR is stuck. @Vexu and @andrewrk don't seem to like my way of implementation and don't accept my explanations.
I have long run out of ideas on how to usefully improve this PR or be more convincing explaining my implementation strategy.

How can we move forward from here? Do we still want this or should we just abandon it to restore peace of mind for all of us?

andrewrk commented:
> How can we move forward from here? Do we still want this or should we just abandon it to restore peace of mind for all of us?

Just leave it, one of us will take it from here. Thanks for your efforts :+1:

andrewrk commented:
Closing this PR because it is not mergeable. The issue #9786 is still open to track the problem, and the code from this PR can be referenced just as easily with it closed.
